cpp/src/arrow/util/int128.cc

```diff
@@ -0,0 +1,537 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <cmath>
+#include <iomanip>
+#include <limits>
+#include <sstream>
+
+#include "arrow/util/int128.h"
+#include "arrow/util/logging.h"
+
+namespace arrow {
+namespace decimal {
+
+static constexpr uint64_t kIntMask = 0xFFFFFFFF;
+static constexpr uint64_t kCarryBit = 1UL << 32UL;
+
+Int128::Int128(const std::string& str) : Int128() {
+  const size_t length = str.length();
+
+  if (length > 0) {
+    bool is_negative = str[0] == '-';
+    auto posn = static_cast<size_t>(is_negative);
+
+    while (posn < length) {
+      const size_t group = std::min(18UL, length - posn);
+      const auto chunk = static_cast<int64_t>(std::stoll(str.substr(posn, group)));
```

Comment:
These exceptions won't be thrown, because we validate the string that goes into this function before it's called. `FromString -> StringToInteger -> Int128 constructor` is the code path, and we're validating that the only thing that can possibly go into this function is decimal digits in `FromString`.